Documentation: adding new components (#16845)
WordPress/gutenberg: packages/components/src/CONTRIBUTING.md

# Contributing components

> These guidelines document how to contribute to components in the [`@wordpress/components`](https://www.npmjs.com/package/@wordpress/components) npm package, or “WordPress Components”. This can be addition, modification, or deprecation.

Reviewer: ```suggestion
You can contribute by adding, modifying, or deprecating components, as well as helping with design, development, and documentation.
```

## Does it belong in the component library?

> A component library should include what’s shared across many products and omit what’s not. Questions to help determine if a component should be added:

Reviewer: ```suggestion
A component library should include components that are generic and flexible enough to work across a variety of products. It should include what’s shared across many products and omit what’s not. 

To determine if a component should be added, ask yourself:
```

> - Is it relevant to other products/plugins?

Reviewer: ```suggestion
- Could this component be used by other products/plugins?
```

> - Is there overlap with any existing components?

Reviewer: ```suggestion
- Does the new component overlap (in functionality or visual design) with any existing components?
```

> - How much will it cost to make and maintain?

Reviewer: The word "cost" here might be conceptually difficult, since the first instinct to think of it in terms of money, rather than time or opportunity costs.

> - Can you articulate how the component should be used?

Reviewer: ```suggestion
- Is there a clear use case requirement for the component, and can you articulate usage guidelines?
```

Side note: I'm not sure about "Can you articulate" here in general—it feels a bit aggressive in terms of putting the onus on the component contributor and could put people off from suggesting components.

Here’s a flowchart that can help determine if a new component is necessary:

> [![New component flowchart](https://wordpress.org/gutenberg/files/2019/07/New_component_flowchart.png)](https://coggle.it/diagram/WtUSrld3uAYZHsn-/t/new-ui-component/992b38cbe685d897b4aec6d0dd93cc4b47c06e0d4484eeb0d7d9a47fb2c48d94)

Reviewer: I love a good flowchart, but it might be worth rebuilding this so it's a bit easier to read and uses some of WordPress' visual styles. It's a bit hard to read, and the horizontal alignment is going to make it hostile to users on mobile devices. (I'm happy to help here or offer more specific guidance as needed.)

Reviewer 2: Thanks Sara, I think that sounds great. I'll consider it a non-blocking issue.

> ## First steps to contributing

Reviewer: ```suggestion
## First steps
```

> Once you’re ready to contribute to WordPress Components, start by opening a GitHub issue. Include a detailed description in which you:

Reviewer: ```suggestion
If you have a component you'd like added or changed, start by opening a GitHub issue. Include a detailed description in which you:
```

- Explain the rationale
- Detail the intended behavior
- Clarify whether it’s a variation of an existing component, or a new asset
- Include mockups of any fidelity (optional)
- Include any inspirations from other products (optional)

> This issue will be the staging ground for the contribution and an opportunity for the community to weigh in with any suggestions. Reviewers can also check if there is overlap with an existing component, or if there is a more appropriate location for the component.

Reviewer: ```suggestion
This issue will be used to discuss the proposed changes and track progress. Reviewers start by discussing the proposal to determine if it's appropriate for WordPress Components. They can also check if there is overlap with an existing component, or if there is a more appropriate location for the component.
```

Reviewer: "a more appropriate location" feels super awkward and really implies physical space, but I can't think of a better term off the top of my head here.

It’s encouraged to surface works-in-progress. If you’re not able to complete all of the parts yourself, someone in the community may be able to pick up where you left off.

> ## Parts of a component contribution

Reviewer: ```suggestion
## Next steps
```

Once the team has discussed and approved the change, it's time to start implementing it.

Reviewer 2: > Once the team has discussed and approved the change, it's time to start implementing it.

Did you mean to include this line in the suggestion? I like it.

Reviewer: I sure did! I was trying to make the steps a bit more sequential & balanced, so "First steps" and only _then_ when that's all done should we move on to "Next steps."

We could also move the "you don't have to do it all yourself" bit up here as well, although I think it works well either before or after the list of steps.

1. **Provide a rationale**: Explain how your component will add value to the system and the greater product ecosystem. Be sure to include any user experience and interaction descriptions.

> 2. **Create a design spec**: Create sizing and styling annotations for all aspects of the component. This spec should provide a developer with everything they need to create the design in code.

Reviewer: An example here would be good.

Reviewer 2: I'll add a link to an example, but we can also provide an image later. Figma pretty much already provides this with their Code view.

> 3. **Create a Figma component**: Any new components or changes to existing components will mirror the [WordPress Components Figma library](https://www.figma.com/file/ZtN5xslEVYgzU7Dd5CxgGZwq/WordPress-Components?node-id=735%3A0), so we’ll need to update the Figma library and publish the changes. Please follow the [guidelines](https://www.figma.com/file/ZtN5xslEVYgzU7Dd5CxgGZwq/WordPress-Components?node-id=746%3A38) for contributing to the Figma libraries.

Reviewer: ```suggestion
3. **Create a Figma component**: Any new components or changes to existing components will be mirrored in the [WordPress Components Figma library](https://www.figma.com/file/ZtN5xslEVYgzU7Dd5CxgGZwq/WordPress-Components?node-id=735%3A0), so we’ll need to update the Figma library and publish the changes. Please follow the [guidelines](https://www.figma.com/file/ZtN5xslEVYgzU7Dd5CxgGZwq/WordPress-Components?node-id=746%3A38) for contributing to the Figma libraries.
```

> 4. **Provide usage documentation**: If the contribution adds additional behavior or expands a component’s features, we’ll need to document them with development and design guidelines. Read through existing component documentation for examples.

Reviewer: What about having the documentation step be #2? Writing documentation, especially for new components, as a first step is a great way of thinking first of usage and application, rather than starting with visuals. Starting with documentation also helps developers and informs that process.

Reviewer: We should also provide a link to a template that can be reused—probably the easiest is a Markdown file stored in the repo, but I'm open to other ideas as well.

> 4. **Provide usage documentation**: If the contribution adds additional behavior or expands a component’s features, we’ll need to document them with development and design guidelines. Read through existing component documentation for examples.

Reviewer: ```suggestion
4. **Draft documentation**: New components need development, design, and accessibility guidelines. Additionally, if your change adds additional behavior or expands a component’s features, those changes will need to be fully documented as well. Read through existing component documentation for examples. Start with a rough draft, and reviewers will help polish documentation.
```

5. **Provide working code**: The component or enhancement must be built in React. See the [developer contribution guidelines](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/develop.md).

> Remember, it’s unlikely that all parts will be done by one person. Someone in the community may be able to help.

Reviewer: ```suggestion
Remember, it’s unlikely that all parts will be done by one person. Contribute where you can, and others will help.
```

## Component refinement

> Before a component is published it will need to be fine-tuned:

Reviewer: This list is a little heavy to read, especially for contributors for whom English is a second language. Can we simplify and provide concrete examples to de-abstractify it a wee bit?

1. **Expand** to a minimally sufficient feature set. Resolve what features should be included in the component.
2. **Reduce** scope to omit contentious features. Review proposed features and prune unessential items lacking consensus.
3. **Quality assurance**: each contribution must be normalized into system standards.

### Quality assurance

To ensure quality, each component should be tested. The testing process should be done during the development of the component and before the component is published. 

> - **Accessibility**: Has the design and implementation accounted for accessibility? Please use the [WordPress accessibility guidelines](https://make.wordpress.org/accessibility/handbook/best-practices/).

Reviewer: We should definitely have an accessibility review be a blocker here. 

```suggestion
- **Accessibility**: Has the design and implementation accounted for accessibility? Please use the [WordPress accessibility guidelines](https://make.wordpress.org/accessibility/handbook/best-practices/) and use the "Needs Accessibility Feedback" tag to get a review from the accessibility team. It's best to request a review early (at the documentation stage) in order to ensure the component is designed inclusively from the outset.
```

> - **Visual quality**: Does the component apply visual style — color, typography, icons, space, borders, and more — using appropriate variables, and does it follow [visual guidelines](https://make.wordpress.org/design/handbook/design-guide/)?

Reviewer: Probably want to make design review a blocker here too.

- **Sufficient states & variations**: Does it cover all the necessary variations (primary, secondary, dense, etc.) and states (default, hover, active, disabled, loading, etc.), given the intended scope?
- **Content resilience**: Is each dynamic word or image element resilient to too much, too little, and no content at all, respectively? How long can labels be, and what happens when you run out of space?
- **Composability**: Does it fit well when placed next to or layered with other components to form a larger composition?
- **Functionality**: Do all behaviors function as expected?
- **Browser support**: Has the component visual quality and accuracy been assessed across Safari, Chrome, Firefox, IE, and other browsers across relevant devices? Please adhere to our [browser support requirements](https://github.com/WordPress/gutenberg/blob/master/packages/browserslist-config/index.js).
- **Documentation**: Ensure that the component has proper documentation for development, design, and accessibility.

## Deprecation

> A component/feature may need deprecation if:

Reviewer 3: We should emphasize that this applies not only to components but also to their props, whenever we add a new one, rename or remove. This should be reflected in the docs but at the same time we should ensure that old code still works as close as possible to what was before and deprecation warnings are outputted to the Web Console.

Reviewer 4: I wonder if this section belongs to this document, since this is generic and applies to any public API of the project. (A backward compatibility policy document)

Reviewer 3: From my experience, it doesn't harm to repeat it a couple of times in our docs :D

Reviewer: I have no idea where our backward compatibility policy document is—linking to it here might be a good approach, but it's definitely helpful being as explicit as possible about our approach here, specifically with reference to how that works for components.

Reviewer 2: I agree there are some component specific things that can be included here.

- There’s a preferred alternative or it’s no longer in demand

> - It’s no longer necessary 

Reviewer: These first two points are basically the same thing and can likely be combined.

Reviewer: What about if a component changes name or we deprecate features or props?

- It’s moved to a different location

At no point does the deprecated component become unavailable. Instead, the commitment to support it stops. The following steps are needed to deprecate a component:

> 1. Communicate intent via regular channels.

Reviewer: This is a bit vague. What are the regular channels?

Reviewer 2: I'll consider these issues on the deprecation section as non-blocking.

    1. Describe reasoning for deprecation.
    2. Decide on a timeline.
2. Ensure backwards compatibility to keep it available.

> 3. Clearly mark the component as deprecated in documentation and other channels.

Reviewer: 💯 

Maybe get a bit more explicit here about the "other channels"—we want to mark as deprecated in docs, in Figma, maybe output a console warning or some such for devs? What else?

> # Contributing components

Reviewer 5: Maybe "Contributing to WordPress Interface Components"?
